AppCenter/AppCenter/Internals/Util/MSEncrypter.m

```diff
@@ -117,52 +65,150 @@ - (NSString *_Nullable)decryptString:(NSString *)string {
+    cipherText = data;
   } else {
-    result = [NSData dataWithBytes:cipherBuffer length:numBytesDecrypted];
-    if (!result) {
-      MSLogWarning([MSAppCenter logTag], @"Could not create NSData object from decrypted bytes.");
+
+    // Extract key from metadata.
+    NSString *keyTag = [metadata componentsSeparatedByString:kMSEncryptionMetadataInternalSeparator][0];
+    NSRange ivRange = NSMakeRange(metadataLocation + 1, kCCBlockSizeAES128);
+
+    // Metadata, separator, and initialization vector.
+    size_t cipherTextPrefixLength = metadataLocation + 1 + kCCBlockSizeAES128;
+    NSRange cipherTextRange = NSMakeRange(cipherTextPrefixLength, [data length] - cipherTextPrefixLength);
+    initializationVector = [data subdataWithRange:ivRange];
+    cipherText = [data subdataWithRange:cipherTextRange];
+    key = [self getKeyWithKeyTag:keyTag];
+  }
+  return [[self class] performCryptoOperation:kCCDecrypt input:cipherText initializationVector:initializationVector key:key];
+}
+
++ (NSString *)getCurrentKeyTag {
+  @synchronized(classLock) {
+    NSString *keyMetadata = [[MSUserDefaults shared] objectForKey:kMSEncryptionKeyMetadataKey];
+    if (!keyMetadata) {
+      [self rotateToNewKeyTag:kMSEncryptionKeyTagAlternate];
+      return kMSEncryptionKeyTagAlternate;
     }
+
+    // Format is {keyTag}/{expiration as iso}.
+    NSArray *keyMetadataComponents = [keyMetadata componentsSeparatedByString:kMSEncryptionMetadataInternalSeparator];
+    NSString *keyTag = keyMetadataComponents[0];
+    NSString *expirationIso = keyMetadataComponents[1];
+    NSDate *expiration = [MSUtility dateFromISO8601:expirationIso];
+    BOOL isNotExpired = [[expiration laterDate:[NSDate date]] isEqualToDate:expiration];
+    if (isNotExpired) {
+      return keyTag;
+    }
+
+    // Key is expired and must be rotated.
+    if ([keyTag isEqualToString:kMSEncryptionKeyTagOriginal]) {
+      keyTag = kMSEncryptionKeyTagAlternate;
+    } else {
+      keyTag = kMSEncryptionKeyTagOriginal;
+    }
+    [self rotateToNewKeyTag:keyTag];
+    return keyTag;
   }
-  free(cipherBuffer);
-  return result;
 }
 
-+ (void)deleteKeyWithTag:(NSString *)keyTag {
-  [MSKeychainUtil deleteStringForKey:keyTag];
++ (void)rotateToNewKeyTag:(NSString *)newKeyTag {
+  NSDate *expiration = [[NSDate date] dateByAddingTimeInterval:kMSEncryptionKeyLifetimeInSeconds];
+  NSString *expirationIso = [MSUtility dateToISO8601:expiration];
+
+  // Format is {keyTag}/{expiration as iso}.
+  NSString *keyMetadata = [@[ newKeyTag, expirationIso ] componentsJoinedByString:kMSEncryptionMetadataInternalSeparator];
+  [[MSUserDefaults shared] setObject:keyMetadata forKey:kMSEncryptionKeyMetadataKey];
+}
+
+- (NSData *)getKeyWithKeyTag:(NSString *)keyTag {
+  NSData *key;
+  @synchronized(self) {
+
+    // Read inside of a synchronized block because NSDictionary is not thread safe.
+    key = [self.keys objectForKey:keyTag];
+  }
+
+  // If key is not cached, try loading it from Keychain.
+  if (!key) {
```

Comment:
Should this be in below synchronized block? Isn't it possible that the key is generated right before the line 154 `if (!stringKey) {`?

Reply:
Yep, it is possible. That's why we re-check on line 158, inside the block. Doing it this way ensures that the synchronized code path is rarely hit.

Reply:
Can't you simplify locks?